### PR TITLE
Mac: Support projection changes that add new files and delete folder

### PR DESCRIPTION
**Pull request by wilbaker**
Changes for #231 and #223 

*Summary of Changes*
- Updated `PrjFS_DeleteFile` so that it can delete files and folders on Mac (#231)
- Updated VFSForGit to track expanded folders
- Updated the projection change code so that after updating file placeholders VFSForGit will re-expand folders that were previously expanded.  VFSForGit does this by checking the contents of each expanded folder in the the projection and ensuring that all of the that folder's children (files and folders) are on disk.

*Folder re-expansion logic*
1. GVFS records what folders have been expanded
2. When a projection change occurs VFSForGit looks up the expanded folders in the projection
3. If there are any children in the projection for the folder it means that their skip-worktree bit is on, and if the skip-worktree bit is on and the file\folder is not in the placeholder list VFSForGit needs to write it out.

### TODOs
- [X] Address new TODOs that have been added in the code
- [X] Enable additional functional tests

```
 Test Run Summary
   Overall result: Warning
   Test Count: 330, Passed: 326, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 4
     Skipped Tests - Ignored: 4, Explicit: 0, Other: 0
   Start time: 2018-09-06 19:21:38Z
     End time: 2018-09-06 19:37:06Z
     Duration: 928.064 seconds
```

### CODE REVIEW

File: GVFS/GVFS.Common/FileSystem/IPlatformFileSystem.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
     public interface IPlatformFileSystem
     {
         bool SupportsFileMode { get; }
+        bool EnumerationExpandsDirectories { get; }
```
Comment (sanoursa) [inline GVFS/GVFS.Common/FileSystem/IPlatformFileSystem.cs]: Should this live on `IKernelDriver`?

File: GVFS/GVFS.Common/GVFSConstants.cs [hunk 2]
```diff
@@ -7,6 +7,7 @@ public static partial class GVFSConstants
         public const int ShaStringLength = 40;
         public const int MaxPath = 260;
         public const string AllZeroSha = "0000000000000000000000000000000000000000";
+        public const string ExpandedFolderSha = "X_EXPANDED_FOLDER00000000000000000000000";
```
Comment (sanoursa) [inline GVFS/GVFS.Common/GVFSConstants.cs]: This is no longer a sha. Does that matter? Could we just pick another fake, but actual, sha value?

File: GVFS/GVFS.Common/GVFSConstants.cs [hunk 3]
```diff
@@ -7,6 +7,7 @@ public static partial class GVFSConstants
         public const int ShaStringLength = 40;
         public const int MaxPath = 260;
         public const string AllZeroSha = "0000000000000000000000000000000000000000";
```
Comment (sanoursa) [inline GVFS/GVFS.Common/GVFSConstants.cs]: Rename this to `EmptyFolderSha`?

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 4]
```diff
@@ -1094,14 +1099,25 @@ private void UpdatePlaceholders()
...
+                    IEnumerable<PlaceholderListDatabase.PlaceholderData> folderPlaceholderData = placeholderListCopy.Where(x => x.IsFolder);
+
+                    HashSet<string> folderPlaceholders = new HashSet<string>(folderPlaceholderData.Select(x => x.Path), StringComparer.OrdinalIgnoreCase);
+                    foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in folderPlaceholderData.OrderByDescending(x => x.Path))
+                    {
+                        if (GVFSPlatform.Instance.FileSystem.EnumerationExpandsDirectories && folderPlaceholder.IsExpandedFolder)
+                        {
+                            this.ReExpandFolder(blobSizesConnection, folderPlaceholder.Path, updatedPlaceholderList, folderPlaceholders, folderPlaceholdersToKeep);
+                        }
+
```
Comment (sanoursa) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Are we missing an else here?

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 5]
```diff
@@ -1094,14 +1099,25 @@ private void UpdatePlaceholders()
...
                 this.blobSizes.Flush();
 
-                // Waiting for the UpdateOrDeleteFilePlaceholder to fill the folderPlaceholdersToKeep before trying to remove folder placeholders
-                // so that we don't try to delete a folder placeholder that has file placeholders and just fails
-                foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in placeholderListCopy.Where(x => x.IsFolder).OrderByDescending(x => x.Path))
-                {
-                    this.TryRemoveFolderPlaceholder(folderPlaceholder, updatedPlaceholderList, folderPlaceholdersToKeep);
+                using (BlobSizes.BlobSizesConnection blobSizesConnection = this.blobSizes.CreateConnection())
+                {
+                    // Waiting for the UpdateOrDeleteFilePlaceholder to fill the folderPlaceholdersToKeep before trying to remove folder placeholders
```
Comment (sanoursa) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Where are we waiting? I don't see it in the code.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 6]
```diff
@@ -1257,17 +1273,107 @@ private string GetNewProjectedShaForPlaceholder(string path)
...
+                }
+                else
+                {
+                    if (!updatedPlaceholderList.ContainsKey(childRelativePath))
+                    {
+                        FileData childFileData = childEntry as FileData;
+                        string sha = childFileData.Sha.ToString();
+
+                        // TODO(Mac): Issue #245, handle failures of WritePlaceholder
+                        this.fileSystemVirtualizer.WritePlaceholder(childRelativePath, childFileData.Size, isDirectory: false, shaContentId: sha);
```
Comment (sanoursa) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Break out the parameters onto separate lines to match the style above?

Comment (sanoursa) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: And actually the contents of this if block and the one above are virtually identical. Is there an easy way to avoid the duplication?

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 7]
```diff
@@ -1257,17 +1273,107 @@ private string GetNewProjectedShaForPlaceholder(string path)
...
+                            childRelativePath,
+                            new PlaceholderListDatabase.PlaceholderData(childRelativePath, sha));
+
+                        folderPlaceholdersToKeep.Add(relativeFolderPath);
+                    }
+                }
+            }
+        }
+
         private void TryRemoveFolderPlaceholder(
```
Comment (sanoursa) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Odd to have a Try method with void return, rather than a bool. Why does the Try mean in this context? Is this method just best-effort?

File: GVFS/GVFS.Common/GVFSConstants.cs (hunk 2 above)
Comment (wilbaker) [reply GVFS/GVFS.Common/GVFSConstants.cs]: I think I picked the wrong spot to put this constant because this is really specific to the placeholder database (and as you mentioned it's not a sha).  I'm thinking we do something like:

- Leave `AllZeroSha` here (as-is)
- Add `EmptyFolderSha` to `PlaceholderListDatabase`
- Move `ExpandedFolderSha` to `PlaceholderListDatabase` and rename it something like `ExpandedFolderValue`

Optionally `AddAndFlush(string path, string sha)` could be renamed to something like `AddAndFlushFile(string path, string sha)` and we could add `AddAndFlushFolder(string path, bool isExpanded)`.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 5 above)
Comment (sanoursa) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: I see now that you just moved this comment, so it's not new. But I'm having a hard time parsing what it means.

Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: The wait happens in `ProcessListOnThreads`:

```
for (int i = 0; i < processThreads.Length; i++)
{
    processThreads[i].Join();
}
```
But I that's not clear from the comment.  I can change it to something like:

```
// Don't attempt to delete folder placeholders until UpdateOrDeleteFilePlaceholder 
// has finished populating folderPlaceholdersToKeep
```

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 4 above)
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: There's no else because we still want to attempt to remove empty placeholders on Mac.

The call to `ReExpandFolder` happens first to (potentially) add more entries to `folderPlaceholdersToKeep`.

Let me know if you think a comment would be useful here.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 6 above)
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: I'll work on cleaning this up

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 7 above)
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Yep it looks like best effort.  I think this can be renamed to something like `RemoveFolderPlaceholderIfEmpty`.

File: GVFS/GVFS.FunctionalTests/Categories.cs [hunk 8]
```diff
@@ -14,8 +14,8 @@ public static class MacTODO
             // The FailsOnBuildAgent category is for tests that pass on dev
             // machines but not on the build agents
             public const string FailsOnBuildAgent = "FailsOnBuildAgent";
-
-            public const string NeedsLockHolder = "NeedsDotCoreLockHolder"; 
+            public const string NeedsLockHolder = "NeedsDotCoreLockHolder";
+            public const string NeedsCachePoisonFix = "NeedsCachePoisonFix";
```
Comment (sanoursa) [inline GVFS/GVFS.FunctionalTests/Categories.cs]: Could you please add this category to `MultithreadedReadWriteTests.CanReadVirtualFileInParallel` and remove the `WindowsOnly` category from it?

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 5 above)
Comment (sanoursa) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: I think I would just remove the comment. My guess is that it was written before the ProcessListOnThreads was factored out, and we were actually waiting on threads to complete right before this comment. Now the comment is just confusing - of course we're waiting for it, this code won't run until that other method returns, and folderPlaceholdersToKeep is an output of that method.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 4 above)
Comment (sanoursa) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Yea a comment would be helpful, I think. And a functional test that fails if we add that else, because it's so very tempting to do so.

File: GVFS/GVFS.Common/GVFSConstants.cs (hunk 2 above)
Comment (wilbaker) [reply GVFS/GVFS.Common/GVFSConstants.cs]: > That all makes sense, but what about the value itself?

The rules for the value:

- Must be 40 characters long (due to validation in `PlaceholderListDatabase.TryParseAddLine`)
- Must not be an actual valid SHA (otherwise we risk a file's SHA matching `ExpandedFolderValue` and VFSForGit incorrectly thinking the entry is for a folder)

We have a lot of freedom within those rules, is there a different string you'd prefer?

File: GVFS/GVFS.FunctionalTests/Tests/GitCommands/CheckoutTests.cs [hunk 9]
```diff
@@ -237,7 +247,7 @@ public void CheckoutBranchAfterReadingAllFilesAndVerifyContentsCorrect()
                 .WithDeepStructure(this.FileSystem, this.ControlGitRepo.RootPath, compareContent: true);
 
             // Verify modified paths contents
-            GVFSHelpers.ModifiedPathsContentsShouldEqual(this.FileSystem, this.Enlistment.DotGVFSRoot, "A .gitattributes" + Environment.NewLine);
+            GVFSHelpers.ModifiedPathsContentsShouldEqual(this.FileSystem, this.Enlistment.DotGVFSRoot, "A .gitattributes\r\n");
```
Comment (wilbaker) [inline GVFS/GVFS.FunctionalTests/Tests/GitCommands/CheckoutTests.cs]: Suggestion from Kevin: Use a named constant for "\r\n" here (and possibly in the the database code as well)

File: GVFS/GVFS.Common/GVFSConstants.cs (hunk 2 above)
Comment (sanoursa) [reply GVFS/GVFS.Common/GVFSConstants.cs]: Isn't `0000000000000000000000000000000000000000` a valid sha? It seems like we'd have just as much chance of hitting `0000000000000000000000000000000000000001` as all-zeros, yea?

Comment (wilbaker) [reply GVFS/GVFS.Common/GVFSConstants.cs]: I believe that when @kewillford first added folders to the placeholder list he confirmed that the all-zero SHA will never be a valid SHA for a git blob.  @kewillford, can you confirm?

(And if it *could*, then I'd argue we're better off adding an upgrade step to convert those entries to a different invalid SHA rather than using 00...0001 for expanded folders)

Comment (sanoursa) [reply GVFS/GVFS.Common/GVFSConstants.cs]: Yep, why have a value here that even looks like a sha, if we want it to _not_ be a sha?

Comment (wilbaker) [reply GVFS/GVFS.Common/GVFSConstants.cs]: > There is a check in the git code when writing out the index that will error out if there is a null sha found in the index unless an environment variable is set. That said, not using something that looks like a sha at all for these seems like a better fit for our case especially now that there is more that one.

@kewillford @sanoursa, any objections to adding a major upgrade step to convert the all-zero SHAs to something that doesn't look like a sha?

File: GVFS/GVFS.FunctionalTests/Tests/GitCommands/GitRepoTests.cs [hunk 10]
```diff
@@ -409,7 +409,7 @@ protected void RunFileDirectoryEnumerateTest(string command, string commandBranc
         protected void RunFileDirectoryReadTest(string command, string commandBranch = DirectoryWithFileAfterBranch)
         {
             this.SetupForFileDirectoryTest(commandBranch);
-            this.FileContentsShouldMatch("file.txt\\file.txt");
+            this.FileContentsShouldMatch("file.txt", "file.txt");
```
[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.FunctionalTests/Tests/GitCommands/GitRepoTests.cs]: Yep it's happening inside `FileContentsShouldMatch`:

```
        protected void FileContentsShouldMatch(params string[] filePathPaths)
        {
            string filePath = Path.Combine(filePathPaths);
```

File: GVFS/GVFS.Tests/Should/EnumerableShouldExtensions.cs [hunk 11]
```diff
@@ -122,6 +122,11 @@ public static IEnumerable<T> ShouldMatchInOrder<T>(this IEnumerable<T> group, IE
             return group;
         }
 
+        public static IEnumerable<T> ShouldMatchInOrder<T>(this IEnumerable<T> group, params T[] expectedValues)
+        {
+            return group.ShouldMatchInOrder(expectedValues, (t1, t2) => t1.Equals(t2));
```
[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.Tests/Should/EnumerableShouldExtensions.cs]: Good idea, I had trouble doing that before, but I think I forgot to cast `expectedValues` to an `IEnumerable`

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 12]
```diff
@@ -1084,7 +1089,7 @@ private void UpdatePlaceholders()
             using (ITracer activity = this.context.Tracer.StartActivity("UpdatePlaceholders", EventLevel.Informational, metadata))
             {
                 ConcurrentHashSet<string> folderPlaceholdersToKeep = new ConcurrentHashSet<string>();
-                ConcurrentBag<PlaceholderListDatabase.PlaceholderData> updatedPlaceholderList = new ConcurrentBag<PlaceholderListDatabase.PlaceholderData>();
+                ConcurrentDictionary<string, PlaceholderListDatabase.PlaceholderData> updatedPlaceholderList = new ConcurrentDictionary<string, PlaceholderListDatabase.PlaceholderData>(StringComparer.Ordinal);
```
[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: > Also how much overhead does adding the string key add? 

Let me measure that.  If it's significant we might need to further split this code out between Mac and Windows (because we do need the dictionary on Mac).

> Should this be a ConcurrentDictionary<string, string> and we build the PlaceholderListDatabase.PlaceholderData when calling WriteAllEntriesAndFlush?

That's a good idea, let me give that a try before making any perf measurements.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 13]
```diff
@@ -1094,14 +1099,25 @@ private void UpdatePlaceholders()
...
-                foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in placeholderListCopy.Where(x => x.IsFolder).OrderByDescending(x => x.Path))
-                {
-                    this.TryRemoveFolderPlaceholder(folderPlaceholder, updatedPlaceholderList, folderPlaceholdersToKeep);
+                using (BlobSizes.BlobSizesConnection blobSizesConnection = this.blobSizes.CreateConnection())
+                {
+                    // Waiting for the UpdateOrDeleteFilePlaceholder to fill the folderPlaceholdersToKeep before trying to remove folder placeholders
+                    // so that we don't try to delete a folder placeholder that has file placeholders and just fails
+                    IEnumerable<PlaceholderListDatabase.PlaceholderData> folderPlaceholderData = placeholderListCopy.Where(x => x.IsFolder);
+
+                    HashSet<string> folderPlaceholders = new HashSet<string>(folderPlaceholderData.Select(x => x.Path), StringComparer.OrdinalIgnoreCase);
```
[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Let me measure that as well.

One question I had looking at this code, do you remember why the folder placeholder list is ordered by descending before calling TryRemoveFolderPlaceholder? 

I assumed it was being done so that child folders would be removed before their parents, does that sound right to you?

>  is now being enumerated twice. Here and the line below

Another option I'll look into is removing this call (higher up in the method):

``` placeholderListCopy.Where(x => !x.IsFolder).ToList(),```

And replacing it with code that will walk `placeholderListCopy` to build both a file list and the `HashSet`.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 14]
```diff
@@ -1289,14 +1395,17 @@ private string GetNewProjectedShaForPlaceholder(string path)
                     metadata.Add("result.RawResult", result.RawResult);
                     metadata.Add("UpdateFailureCause", failureReason.ToString());                    
                     this.context.Tracer.RelatedEvent(EventLevel.Informational, nameof(this.TryRemoveFolderPlaceholder) + "_DeleteFileFailure", metadata);
+
+                    updatedPlaceholderList.TryAdd(placeholder.Path, placeholder);
```
[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: I'm glad you mentioned this, the idea was that VFSForGit should still track the placeholder if it fails to delete it, but I think I'll replace this line with a comment that references #245 and update #245 to include handling failures to delete folder placeholders.

I think that on Mac if we fail to delete the placeholder folder we should add its path to ModifiedPaths.dat and show the user a message (otherwise the folder will essentially be abandoned and never be re-expanded or deleted again).

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (kewillford) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Yes, we have to make sure all the children are removed first so that we can remove the folder when it is empty since we are only removing individual folders and are depending on getting the directory not empty error if we can't delete because of existing files or folders.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: (I will dig more into our failure options when working on #245)

File: GVFS/GVFS.Common/GVFSConstants.cs (hunk 2 above)
Comment (sanoursa) [reply GVFS/GVFS.Common/GVFSConstants.cs]: I think that's a good idea. And I would request that these special values that don't look like they are part of the path when manually looking at the PlaceholderList file. (i.e. put some delimiter char(s) at the beginning of the value)

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (sanoursa) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Do we have tests that fail if we drop that ordering? This is a very subtle detail that is easy to lose track of over time.

File: GVFS/GVFS.Common/GVFSConstants.cs (hunk 2 above)
Comment (wilbaker) [reply GVFS/GVFS.Common/GVFSConstants.cs]: > I think that's a good idea. And I would request that these special values that don't look like they are part of the path when manually looking at the PlaceholderList file. (i.e. put some delimiter char(s) at the beginning of the value)

Sounds good, I'll start the values with `@`

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: > Do we have tests that fail if we drop that ordering? This is a very subtle detail that is easy to lose track of over time.

I'll try taking that out and see if any tests fail.  The tricky part is even if the ordering were to be removed, I'm not sure that we can come up with a test that would be guaranteed to catch it (as whatever random ordering VFSForGit would use in the future might just happen to work with the test).

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: > Should this be a ConcurrentDictionary<string, string> and we build the PlaceholderListDatabase.PlaceholderData when calling WriteAllEntriesAndFlush

I'm actually going to start with the perf overhead (starting that testing this morning) and see how things look before and after this change.  We're using the same string instances in both the `ConcurrentDictionary` and the `PlaceholderData` and so the memory overhead is just an extra reference.

[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: > Should this be a ConcurrentDictionary<string, string> and we build the PlaceholderListDatabase.PlaceholderData when calling WriteAllEntriesAndFlush?

Also, to confirm, what was the main concern here? Perf or memory usage?

**Comment by wilbaker** [comment]
@sanoursa @kewillford 

Here are the initial numbers comparing the total time of `UpdatePlaceholders`:

Results | Baseline | With Changes
------------ | ------------- | -------------
Average | 4056ms | 4735ms
Min | 3436ms | 2985ms
Max | 4619ms | 5767ms
StdDev | 354ms | 866ms
Count | 11 | 11

We are noticabley worse with the changes, and so I'll keep digging into where the time is going.

*Details*

- The time reported is the time of the full `UpdatePlaceholders` procedure
- File placeholders: 404,130
- Folder placeholders: 60,612
- Ran `git checkout` to move between two branches with the same commit (and so no placeholder would need to be updated on disk)

**Comment by wilbaker** [comment]
@sanoursa @kewillford I just pushed a commit to get our times back down when running on Windows:

Results | Latest Commit
------------ | -------------
Average | 4019ms
Min | 2944ms
Max | 4935ms
StdDev | 638ms
Count | 11

File: GVFS/GVFS.Common/PlaceholderListDatabase.cs [hunk 15]
```diff
@@ -218,18 +278,29 @@ private bool TryParseRemoveLine(string line, out string key, out string error)
...
 
             public string Path { get; }
             public string Sha { get; }
 
             public bool IsFolder
             {
-                get { return this.Sha == GVFSConstants.AllZeroSha; }
+                get 
+                {
+                    return this.Sha.StartsWith(FolderValuePrefix, StringComparison.Ordinal);
```
Comment (sanoursa) [inline GVFS/GVFS.Common/PlaceholderListDatabase.cs]: Why not just compare the full strings? Are the prefixes creating a significant perf win?

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 16]
```diff
@@ -1078,30 +1087,85 @@ private void UpdatePlaceholders()
...
 
-                // Waiting for the UpdateOrDeleteFilePlaceholder to fill the folderPlaceholdersToKeep before trying to remove folder placeholders
-                // so that we don't try to delete a folder placeholder that has file placeholders and just fails
-                foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in placeholderListCopy.Where(x => x.IsFolder).OrderByDescending(x => x.Path))
+                using (BlobSizes.BlobSizesConnection blobSizesConnection = this.blobSizes.CreateConnection())
+                {
+                    // A hash of the folder placeholders is only required if the platform expands directories
+                    HashSet<string> folderPlaceholders = 
+                        GVFSPlatform.Instance.KernelDriver.EnumerationExpandsDirectories ?
+                                    new HashSet<string>(placeholderFoldersListCopy.Select(x => x.Path), StringComparer.OrdinalIgnoreCase) :
```
Comment (sanoursa) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Nit: fix the spacing

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 17]
```diff
@@ -1078,30 +1087,85 @@ private void UpdatePlaceholders()
...
+                int numThreads = Math.Max(8, Environment.ProcessorCount);
+                numThreads = Math.Min(numThreads, placeholderFilesListCopy.Count / minItemsPerThread);
+                numThreads = Math.Max(numThreads, 1);
+
                 ConcurrentHashSet<string> folderPlaceholdersToKeep = new ConcurrentHashSet<string>();
-                ConcurrentBag<PlaceholderListDatabase.PlaceholderData> updatedPlaceholderList = new ConcurrentBag<PlaceholderListDatabase.PlaceholderData>();
+
+                // On platforms that expand on enumeration we need to keep the updated placeholders in a collection
+                // that supports fast searching (for ReExpandFolder)
+                ConcurrentDictionary<string, PlaceholderListDatabase.PlaceholderData> updatedPlaceholderDictionary;
```
Comment (sanoursa) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: It's a little concerning that we're keeping two data structures for the same data. How do we guarantee that they stay in sync?

Comment (sanoursa) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Or is the requirement that it be one or the other? In that case, I'd add a few strategic checks that enforce that and throw an exception if it's violated.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 18]
```diff
@@ -1078,30 +1087,85 @@ private void UpdatePlaceholders()
...
-                foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in placeholderListCopy.Where(x => x.IsFolder).OrderByDescending(x => x.Path))
+                using (BlobSizes.BlobSizesConnection blobSizesConnection = this.blobSizes.CreateConnection())
+                {
+                    // A hash of the folder placeholders is only required if the platform expands directories
+                    HashSet<string> folderPlaceholders = 
+                        GVFSPlatform.Instance.KernelDriver.EnumerationExpandsDirectories ?
+                                    new HashSet<string>(placeholderFoldersListCopy.Select(x => x.Path), StringComparer.OrdinalIgnoreCase) : 
+                                    null;
+                    
+                    foreach (PlaceholderListDatabase.PlaceholderData folderPlaceholder in placeholderFoldersListCopy.OrderByDescending(x => x.Path))
```
Comment (sanoursa) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Is this `OrderByDescending` here to ensure that we walk through the tree from the bottom up? I would add some comments to clarify that.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 19]
```diff
@@ -1257,15 +1319,104 @@ private string GetNewProjectedShaForPlaceholder(string path)
...
+            HashSet<string> existingFolderPlaceholders)
+        {
+            FolderData folderData;
+            if (!this.TryGetOrAddFolderDataFromCache(relativeFolderPath, out folderData))
+            {
+                // Folder is no longer in the projection
+                return;
+            }
+
+            folderData.PopulateSizes(
```
Comment (sanoursa) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Aren't we already gathering the necessary sizes elsewhere? How do we ensure that we're not doing lots of repeated round trips to the server here?

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 20]
```diff
@@ -1257,15 +1319,104 @@ private string GetNewProjectedShaForPlaceholder(string path)
...
-                updatedPlaceholderList.Add(placeholder);
-                return;
+                addPlaceholderToUpdatedPlaceholders(placeholder);
+                return false;
+            }
+
+            if (GVFSPlatform.Instance.KernelDriver.EnumerationExpandsDirectories)
+            {
+                // If enumeration expands directories we should leave folder placeholders
+                // that are still in the projection on disk (even if they are empty).
```
Comment (sanoursa) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Why do we leave them even if they are empty? How then do empty directories after a checkout get cleaned up?

File: ProjFS.Mac/PrjFSLib/PrjFSLib.cpp [hunk 21]
```diff
@@ -411,7 +411,7 @@ PrjFS_Result PrjFS_DeleteFile(
     
     char fullPath[PrjFSMaxPath];
     CombinePaths(s_virtualizationRootFullPath.c_str(), relativePath, fullPath);
-    if (0 != unlink(fullPath))
+    if (0 != remove(fullPath))
```
Comment (sanoursa) [inline ProjFS.Mac/PrjFSLib/PrjFSLib.cpp]: Why was this changed?

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 20 above)
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: > Why do we leave them even if they are empty?

I think should update this comment.  "Empty" here means that on the physical disk the folder is empty (e.g. because the folder has not been expanded yet).

> How then do empty directories after a checkout get cleaned up?

If they are empty from git's perspective they will *not* be in the projection, and so `TryGetOrAddFolderDataFromCache` will return false and VFSForGit will proceed with the delete of the folder.

(On Windows we can't make this "in projection" check because we also have to worry about tombstones that might be on disk)

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 19 above)
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Filed #255 for this

File: GVFS/GVFS.FunctionalTests.Windows/Windows/Tests/DiskLayoutUpgradeTests.cs [hunk 22]
```diff
@@ -348,16 +378,16 @@ private string[] GetPlaceholderDatabaseLinesBeforeUpgrade(string placeholderData
...
-            lines.ShouldContain(x => x == "A GVFS\\GVFS.Common\0" + TestConstants.AllZeroSha);
-            lines.ShouldContain(x => x == "A GVFS\\GVFS.Common\\Git\0" + TestConstants.AllZeroSha);
-            lines.ShouldContain(x => x == "A GVFS\\GVFS.Tests\0" + TestConstants.AllZeroSha);
-            lines.ShouldContain(x => x == "A GVFS\\GVFS.Tests\\Properties\0" + TestConstants.AllZeroSha);
+            lines.ShouldContain(x => x == "A Scripts\0" + TestConstants.PartialFolderPlaceholderDatabaseValue);
+            lines.ShouldContain(x => x == "A GVFS\0" + TestConstants.PartialFolderPlaceholderDatabaseValue);
+            lines.ShouldContain(x => x == "A GVFS\\GVFS.Common\0" + TestConstants.PartialFolderPlaceholderDatabaseValue);
+            lines.ShouldContain(x => x == "A GVFS\\GVFS.Common\\Git\0" + TestConstants.PartialFolderPlaceholderDatabaseValue);
+            lines.ShouldContain(x => x == "A GVFS\\GVFS.Tests\0" + TestConstants.PartialFolderPlaceholderDatabaseValue);
+            lines.ShouldContain(x => x == "A GVFS\\GVFS.Tests\\Properties\0" + TestConstants.PartialFolderPlaceholderDatabaseValue);
```
Comment (kewillford) [inline GVFS/GVFS.FunctionalTests.Windows/Windows/Tests/DiskLayoutUpgradeTests.cs]: Is there some way to group some of these so we don't have to repeat 2 more times?

File: GVFS/GVFS.Virtualization/FileSystem/FileSystemVirtualizer.cs [hunk 23]
```diff
@@ -13,6 +13,8 @@ public abstract class FileSystemVirtualizer : IDisposable
     {
         public const byte PlaceholderVersion = 1;
 
+        protected static readonly byte[] FolderContentId = Encoding.Unicode.GetBytes(GVFSConstants.AllZeroSha);
```
Comment (kewillford) [inline GVFS/GVFS.Virtualization/FileSystem/FileSystemVirtualizer.cs]: Why did this say as `AllZeroSha`?

Comment (wilbaker) [reply GVFS/GVFS.Virtualization/FileSystem/FileSystemVirtualizer.cs]: This is what's actually stored on disk in the folder placeholders.  I wanted to avoid changing the on-disk format of the folder placeholders themselves.

File: GVFS/GVFS.FunctionalTests.Windows/Windows/Tests/DiskLayoutUpgradeTests.cs (hunk 22 above)
Comment (wilbaker) [reply GVFS/GVFS.FunctionalTests.Windows/Windows/Tests/DiskLayoutUpgradeTests.cs]: That's a good idea, let me see how I can clean these up.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 24]
```diff
@@ -1374,7 +1554,7 @@ private string GetNewProjectedShaForPlaceholder(string path)
             PlaceholderListDatabase.PlaceholderData placeholder,
```
Comment (kewillford) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Not for this PR but the method name `ProcessGvUpdateDeletePlaceholderResult` needs to change at some point.

File: GVFS/GVFS.Virtualization/FileSystem/FileSystemVirtualizer.cs (hunk 23 above)
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/FileSystem/FileSystemVirtualizer.cs]: > So we never check this value and don't expect to need to tell the difference using ContentId whether a folder is partial or expanded?

That's right.  We're currently only checking the contentID of files (when we get a hydration request).

On Mac the kext uses the`FileFlags_IsEmpty` flag to check if the directory has been expanded or not.

```if (FileFlagsBitIsSet(currentVnodeFileFlags, FileFlags_IsEmpty))```

I'd rather not change this now unless we have to (to leave things cleaner for ourselves if we do need to change it some day).

Comment (kewillford) [reply GVFS/GVFS.Virtualization/FileSystem/FileSystemVirtualizer.cs]: Sounds good.  Thanks!
